### ISSUE TITLE
Proposed performance enhancement for hub method resolution/dispatching

### ISSUE THREAD

**Issue by drub0y** [SignalR/SignalR#351]
# TL;DR

We can gain 3.6% performance in hub method resolution by introducing a secondary cache in `ReflectedMethodDescriptorProvider::TryGetHubMethod` which is in the direct line of hub method dispatching.
# Detailed Explanation

After some discussion on JabbR I created a test where I make 20k sequential calls to the `DrawingPad` sample hub's `DrawLine` method from a test C# client so that I could profile for any hotspots in the SignalR codebase that deals with hub method dispatching since it is going to be some of the most frequently executed code. In profiling this scenario only one particular SignalR method stood out in the reports "Functions With Most Inidvidual Work" section and that was `ReflectedMethodDescriptorProvider::TryGetHubMethod` with an Application Exclusive Time % of 3.9.

As you may know, every call to a hub method with the default dependency chain results in the following call chain: `HubDispatcher::OnReceivedAsync`->`DefaultHubManager::GetHubMethod`->`ReflectedMethodDescriptorProvider::TryGetHubMethod`. Today, `TryGetHubMethod` always resolves the method using some logic that, when we're talking about this kind of volume, is rather expensive. The logic, which involves a LINQ query, List<MethodDescriptor> allocation and a .Count check, is necessary because `ReflectedMethodDescriptorProvider` has to take into account multiple methods with the same name/parameter names and, if multiple are found, none are selected and null is returned instead. This behavior is desired as opposed to selecting one of the methods at random and just allowing it to be the method that handles the call for the hub.  At the same time the `ReflectedMethodDescriptorProvider` returns the multiple methods from it's `GetMethods` implementation and in order to continue to fulfill that behavior we cannot simply exclude methods that are overloaded from the `BuildMethodCacheFor` implementation. This then leaves a bit of a hole because, while the `MethodDescriptors` _are_ cached, they are cached as a set and the exact method descriptor to fulfill a particular method _is not_ cached.

I am proposing that we introduce a secondary layer of caching in `TryGetMethod` which will use another cache to store the "final" resolved method. The logic will check the cache first and, if the method is not yet cached, it will execute today's existing logic to resolve it and finally, assuming a `MethodDescriptor` is successfully resolved, store it in its cache for future lookups to resolve directly. Using my afforementioned test, this proposed implementation now yields an Exclusive Time % of only 0.3 or a 3.6% gain in performance.

I will be submitting a PR with this implementation shortly and will link it to this issue.
# Related Consideration

Finally, I should point out that if we could do away with the need for "soft"-supporting multiple overloads with the same method name we could do away with the cost of a secondary cache. To do this we could either:

  **A.** Modify `ReflectedMethodDescriptorProvider::BuildMethodCacheFor` to fail hard/fast with an exception explaining that there is an unsupported overload situation

   _OR_

  **B.** Simply never cache the methods that are not distinct which would ultimately result in the same runtime exception as today where `HubDispatcher::OnReceivedAsync` will throw an `InvalidOperationException` explaning that the method could not be resolved. 

I would personally _prefer_ to take one of these approaches, but it breaks tests that exist today that validate the method overload scenario and check that the methods are still returned from `ReflectedMethodDescriptorProvider::GetMethods`. I truly question if this test based requirement is real enough to warrant the runtime costs associated with it considering the methods cannot be executed anyway, but I'll leave that up for consideration/discussion because perhaps I am missing something that makes this behavior desireable.
